datafusion/sql/src/interval.rs

```diff
@@ -0,0 +1,214 @@
+                value
+            )));
+        }
+
+        result_days += diff_days as i64;
+
+        if result_days > (i32::MAX as i64) {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Interval field value out of range: {:?}",
+                value
+            )));
+        }
+
+        result_millis += diff_millis as i64;
+
+        if result_millis > (i32::MAX as i64) {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Interval field value out of range: {:?}",
+                value
+            )));
+        }
+    }
+
+    // Interval is tricky thing
+    // 1 day is not 24 hours because timezones, 1 year != 365/364! 30 days != 1 month
+    // The true way to store and calculate intervals is to store it as it defined
+    // It's why we there are 3 different interval types in Arrow
+    if result_month != 0 && (result_days != 0 || result_millis != 0) {
+        let result: i128 = ((result_month as i128) << 96)
+                | ((result_days as i128) << 64)
+                // IntervalMonthDayNano uses nanos, but IntervalDayTime uses milles
+                | ((result_millis * 1_000_000_i64) as i128);
+
+        return Ok(ScalarValue::IntervalMonthDayNano(Some(result)));
+    }
+
+    // Month interval
+    if result_month != 0 {
+        return Ok(ScalarValue::IntervalYearMonth(Some(result_month as i32)));
+    }
+
+    let result: i64 = (result_days << 32) | result_millis;
+    Ok(ScalarValue::IntervalDayTime(Some(result)))
+}
+
+#[cfg(test)]
+mod test {
+    use crate::assert_contains;
+
+    use super::*;
+
+    #[test]
+    fn test_parse_ym() {
+        assert_eq!(
+            parse_interval("months", "1 month").unwrap(),
+            ScalarValue::new_interval_ym(0, 1)
+        );
+
+        assert_eq!(
+            parse_interval("months", "2 month").unwrap(),
+            ScalarValue::new_interval_ym(0, 2)
+        );
+
+        assert_eq!(
+            parse_interval("months", "3 year 1 month").unwrap(),
+            ScalarValue::new_interval_ym(3, 1)
+        );
+
+        assert_contains!(
+            parse_interval("months", "1 years 1 month")
+                .unwrap_err()
+                .to_string(),
+            r#"Invalid input syntax for type interval: "1 years 1 month""#
+        );
+    }
+
+    #[test]
+    fn test_dt() {
+        assert_eq!(
+            parse_interval("months", "5 days").unwrap(),
+            ScalarValue::new_interval_dt(5, 0)
+        );
+
+        assert_eq!(
+            parse_interval("months", "7 days 3 hours").unwrap(),
+            ScalarValue::new_interval_dt(
+                7,
+                (3.0 * SECONDS_PER_HOUR * MILLIS_PER_SECOND) as i32
+            )
+        );
+
+        assert_eq!(
+            parse_interval("months", "7 days 5 minutes").unwrap(),
+            ScalarValue::new_interval_dt(7, (5.0 * 60.0 * MILLIS_PER_SECOND) as i32)
+        );
+    }
+
+    #[test]
+    // TODO file a ticket to track
```

Comment:
I am not quite sure what to make of this -- my current interpretation is that `IntervalMonthDayNano` and the interval parsing logic are inconsistent but I wanted to double check

cc @avantgardnerio

Reply:
When I was doing my implementation, `MonthDayNano` was the one I could never figure out how to test, because I could not get a SQL query to cause one to be generated. Clearly I was unaware of the `parse_interval()` function. I inferred the order of the values based on github conversations, documentation, and how they were ordered in `IntervalDayTimeType`.

So it's possible that `new_interval_mdn()` is not correct in its assumption:
```
        let m = months as u128 & u32::MAX as u128;
        let d = (days as u128 & u32::MAX as u128) << 32;
        let n = (nanos as u128) << 64;
```

I'll pull this branch down and see if I can add more clarity.

Reply:
Looks like `parse_interval()` sees it as:
```
    if result_month != 0 && (result_days != 0 || result_millis != 0) {
        let result: i128 = ((result_month as i128) << 96)
                | ((result_days as i128) << 64)
                // IntervalMonthDayNano uses nanos, but IntervalDayTime uses milles
                | ((result_millis * 1_000_000_i64) as i128);
```

Reply:
It is also entirely possible the but is in  `parse_interval` -- I am not sure anyone uses the `IntervalMonthDayNano` yet -- it was added relatively recently as I recall

Reply:
Relevant conversation:
https://github.com/apache/arrow-datafusion/pull/2797/files#r910183035
https://github.com/apache/arrow-datafusion/pull/2797/files#r910186338

Reply:
I think I found the C impl, and I think `parse_interval()` looks correct https://github.com/apache/arrow/blob/02c8598d264c839a5b5cf3109bfd406f3b8a6ba5/cpp/src/arrow/type.h#L1475

I can make a PR to fix shortly if we're sure this is the correct format. It would be good to have an integration test between C & Rust. Do we have anything like that now?

Reply:
Linking to https://github.com/apache/arrow-rs/pull/2235 for resolution.

Reply:
```suggestion
    // https://github.com/apache/arrow-rs/pull/2235
```